scr/sysbox

```diff
@@ -30,49 +30,163 @@ function retry() {
-
-# Increase default keyring resources to meet sys container demands.
-# For a k8s cluster:
-# keys = 35 + (workers * 23) + (2 * pods)
-# maxbytes = 20 bytes * maxkeys
-sysctl -w kernel.keys.maxkeys=20000
-sysctl -w kernel.keys.maxbytes=$((20*20000))
+# Ensure that kernel-modules expected by system-level apps (running within sys
+# containers) are properly loaded.
+load_required_modules() {
 
-#
-# Start Sysbox components
-#
-mkdir -p /var/lib/sysboxfs
-sysbox-mgr --log /dev/stdout > /var/log/sysbox-mgr.log 2>&1 &
+  # Sysbox requires 'configfs' module to ensure proper operation of containarized
+  # apps requiring access to kernel's config file (e.g. kubeadm).
+  if ! modprobe configfs &> /dev/null; then
+      echo "Could not load configfs kernel module. Exiting ..."
+      return 1
+  fi
+
+  return 0
+}
+
+# Returns linux distro running in the system.
+function get_host_distro() {
+
+  local distro=$(cat /etc/os-release | awk -F"=" '/^ID=/ {print $2}' | tr -d '"')
+
+  echo $distro
+}
+
+# Ensures unprivileged user-ns are allowed.
+function setup_userns() {
+
+  local distro=$1
+
+  if [[ "${distro}" == "ubuntu" ]] ||
+     [[ "${distro}" == "debian" ]]; then
+      echo "1" > /proc/sys/kernel/unprivileged_userns_clone
+
+  elif [[ "${distro}" == "centos" ]] ||
+       [[ "${distro}" == "fedora" ]] ||
+       [[ "${distro}" == "redhat" ]]; then
+
+      # Setting 10k as max user-ns value.
+      max_user_ns=$(</proc/sys/user/max_user_namespaces)
+      if [[ $max_user_ns =~ ^[0-9]+$ ]] && [[ $max_user_ns -lt 10000 ]]; then
+	  echo "10000" > /proc/sys/user/max_user_namespaces
+      fi
+
+  else
+      echo "Unsupported Linux distribution: $distro. Sysbox may not opperate as expected."
+  fi
+}
+
+# Identifies kernel-header's expected path based on distro.
+function kernel_header_path() {
+
+  local distro=$1
+  local path
+
+  if [[ "${distro}" == "centos" ]] ||
+     [[ "${distro}" == "fedora" ]] ||
+     [[ "${distro}" == "redhat" ]]; then
+      path="/usr/src/kernels/$(uname -r)"
+  else
+      path="/usr/src/linux-headers-$(uname -r)"
+  fi
+
+  echo "$path"
+}
+
+# Verifies that a kernel configuration file is found, and if that's not the case,
+# copy it from the original "/boot" folder. Notice that this may not work when
+# running sysbox within a test-priv container, as "/boot" folder may not be around;
+# in those cases initialization will proceed as normal and a log will be dumped to
+# alert the user.
+function setup_kernel_config() {
+
+  local distro=$1
 
-RES=$(retry 10 1 grep -q Ready /var/log/sysbox-mgr.log)
-if [ $? -ne 0 ]; then
-  printf "\nsysbox-mgr failed to start. Here is the log file:\n"
-  cat /var/log/sysbox-mgr.log
-  exit 1
-fi
+  local kernel_path=$(kernel_header_path $distro)
 
+  if [[ ! -f "${kernel_path}"/.config ]]; then
+      # Attempt to find kernel config in /boot path.
+      if [[ -d /boot ]] && [[ -f /boot/config-$(uname -r) ]]; then
+	  cp /boot/config-$(uname -r) "${kernel_path}"/.config
```

Comment:
Porque es necesario hacer esto? Que esta fallando si no lo hacemos? Yo tengo entendido que el kernel config file normalmente reside en /boot/config-*.   Me parece extrano tener que copiarlo al kernel_header_path.

Reply:
Don't see any other solution for this. Without it kubeadm (or any other app looking for kernel config) will complain if this one is not found in the expected path.

Reply:
I am wondering if rather than making a copy at host level, we should instead have sysbox create a bind-mount from the host's `/boot/config-$(uname -r)` to the sys container's `${kernel_path}/.config`

Reply:
As discussed, will push code as it is today as we're mainly dealing with a corner-case scenario (kernel config file not present in headers path after performing kernel upgrade), that appears to be affecting only Debian-Buster.

And yes, sysbox installer will be updated too.